File: lab3/gr9sub2/bosko_anton/src/pstat.py
#!/usr/bin/env python3
import os, sys, re, subprocess, argparse, shutil

def read_file(path):
    try:
        with open(path, 'r', encoding='utf-8', errors='replace') as f:
            return f.read()
    except Exception:
        return None

# ---- parsers (как раньше) ----
def parse_stat(pid):
    s = read_file(f'/proc/{pid}/stat')
    if not s:
        return None
    i1 = s.find('(')
    i2 = s.rfind(')')
    pre = s[:i1].strip().split()
    comm = s[i1+1:i2]
    post = s[i2+1:].strip().split()
    fields = pre + [comm] + post
    utime = int(fields[13])
    stime = int(fields[14])
    ppid = int(fields[3])
    state = fields[2]
    num_threads = int(fields[19])
    rss_pages = int(fields[23])
    return dict(utime=utime, stime=stime, ppid=ppid, state=state,
                num_threads=num_threads, rss_pages=rss_pages, comm=comm)

def parse_status(pid):
    s = read_file(f'/proc/{pid}/status')
    out = {}
    if not s:
        return out
    for line in s.splitlines():
        if ':' not in line: continue
        k,v = line.split(':',1)
        out[k.strip()] = v.strip()
    return out

def parse_io(pid):
    s = read_file(f'/proc/{pid}/io')
    out = {}
    if not s:
        return out
    for line in s.splitlines():
        if ':' not in line: continue
        k,v = line.split(':',1)
        out[k.strip()] = v.strip()
    return out

def try_smaps_rollup(pid):
    s = read_file(f'/proc/{pid}/smaps_rollup')
    if not s:
        return None
    out = {}
    for line in s.splitlines():
        if ':' not in line: continue
        k,v = line.split(':',1)
        out[k.strip()] = v.strip()
    return out

def parse_smaps(pid):
    s = read_file(f'/proc/{pid}/smaps')
    if not s:
        return None
    anon = 0
    filebacked = 0
    cur_is_file = False
    for line in s.splitlines():
        m = re.match(r'^[0-9a-fA-F]+-[0-9a-fA-F]+', line)
        if m:
            parts = line.split()
            if len(parts) >= 6:
                pathname = parts[5]
                cur_is_file = not pathname.startswith('[')
            else:
                cur_is_file = False
            continue
        m = re.match(r'Rss:\s+(\d+)\s*kB', line)
        if m:
            val = int(m.group(1))
            if cur_is_file:
                filebacked += val
            else:
                anon += val
    return {'RssAnon_kB': anon, 'RssFile_kB': filebacked, 'Rss_total_kB': anon+filebacked}

def fmt(bytes_or_kb, kind='bytes'):
    if bytes_or_kb is None:
        return 'N/A'
    if kind == 'kB':
        b = int(bytes_or_kb) * 1024
    else:
        b = int(bytes_or_kb)
    if b >= 1024**3:
        return f'{b/1024**3:.2f} GiB'
    if b >= 1024**2:
        return f'{b/1024**2:.2f} MiB'
    if b >= 1024:
        return f'{b/1024:.1f} KiB'
    return f'{b} B'

# ---- helpers to run system commands for compare ----
def run_cmd(cmd):
    try:
        out = subprocess.check_output(cmd, stderr=subprocess.STDOUT, shell=True, text=True)
        return out
    except subprocess.CalledProcessError as e:
        return f"(command failed) {e}\n{e.output}"

def cmd_exists(name):
    return shutil.which(name) is not None

def compare_with_tools(pid):
    blocks = []
    # ps
    blocks.append(("ps", run_cmd(f"ps -p {pid} -o pid,ppid,comm,state,time,rss,vsz --no-headers")))
    # top
    blocks.append(("top", run_cmd(f"top -b -n1 -p {pid} 2>/dev/null | sed -n '7,9p'")))
    # pidstat: try if exists
    if cmd_exists("pidstat"):
        blocks.append(("pidstat", run_cmd(f"pidstat -p {pid} 1 1")))
    else:
        blocks.append(("pidstat", "(pidstat not found on system)"))
    # /proc raw
    blocks.append(("/proc/<pid>/stat", run_cmd(f"cat /proc/{pid}/stat 2>/dev/null || echo '(no access)'")))
    blocks.append(("/proc/<pid>/status", run_cmd(f"grep -E 'VmRSS|VmSize|Threads|voluntary_ctxt_switches|nonvoluntary_ctxt_switches' /proc/{pid}/status 2>/dev/null || echo '(no access)'")))
    blocks.append(("/proc/<pid>/io", run_cmd(f"cat /proc/{pid}/io 2>/dev/null || echo '(no access)'")))
    return blocks

# ---- main ----
def print_summary(pid, do_compare):
    stat = parse_stat(pid)
    if not stat:
        print("Cannot read /proc/{}/stat".format(pid))
        sys.exit(2)
    status = parse_status(pid)
    io = parse_io(pid)
    smap_roll = try_smaps_rollup(pid)
    smap = None
    if smap_roll:
        rss_kb = None
        rss_anon = None
        rss_file = None
        if 'Rss' in smap_roll:
            try:
                rss_kb = int(smap_roll['Rss'].split()[0])
            except:
                pass
        for key in ['Anon', 'AnonPages', 'RssAnon', 'AnonHugePages']:
            if key in smap_roll:
                try:
                    rss_anon = int(smap_roll[key].split()[0])
                except:
                    pass
        for key in ['File', 'RssFile']:
            if key in smap_roll:
                try:
                    rss_file = int(smap_roll[key].split()[0])
                except:
                    pass
        smap = {'Rss_kB': rss_kb, 'RssAnon_kB': rss_anon, 'RssFile_kB': rss_file}
    else:
        smap = parse_smaps(pid)

    clk_tck = os.sysconf('SC_CLK_TCK')
    page_size = os.sysconf('SC_PAGE_SIZE')
    cpu_time_sec = (stat['utime'] + stat['stime']) / clk_tck
    rss_bytes_from_stat = stat['rss_pages'] * page_size

    print(f"PID: {pid} [{stat['comm']}]")
    print(f"State: {stat['state']}   PPid: {stat['ppid']}   Threads: {stat['num_threads']}")
    print(f"utime (ticks): {stat['utime']}, stime (ticks): {stat['stime']}")
    print(f"CPU time (sec) = (utime+stime)/HZ = ({stat['utime']}+{stat['stime']})/{clk_tck} = {cpu_time_sec:.3f} s")
    print("")
    vmrss = None
    if 'VmRSS' in status:
        m = re.match(r'(\d+)\s*kB', status['VmRSS'])
        if m:
            vmrss = int(m.group(1))
    print(f"VmRSS (status): {vmrss} kB   (formatted: {fmt(vmrss, 'kB')})")
    print(f"rss (stat * pagesize): {rss_bytes_from_stat} bytes ({fmt(rss_bytes_from_stat,'bytes')})")
    if smap:
        print("")
        print("smaps summary (kB):")
        for k,v in smap.items():
            print(f"  {k} : {v}")
        if smap.get('Rss_kB'):
            print(f"  Total Rss from smaps_rollup: {smap['Rss_kB']} kB ({fmt(smap['Rss_kB'],'kB')})")
    if io:
        rb = io.get('read_bytes')
        wb = io.get('write_bytes')
        print("")
        print(f"IO: read_bytes={rb}   write_bytes={wb}")
        try:
            print(f"    formatted: read {fmt(int(rb),'bytes')}  write {fmt(int(wb),'bytes')}")
        except:
            pass
    vcs = status.get('voluntary_ctxt_switches', 'N/A')
    nvcs = status.get('nonvoluntary_ctxt_switches', 'N/A')
    print("")
    print(f"voluntary_ctxt_switches: {vcs}")
    print(f"nonvoluntary_ctxt_switches: {nvcs}")

    if do_compare:
        print("\n--- comparison with system tools ---\n")
        blocks = compare_with_tools(pid)
        for title, out in blocks:
            print("=== ", title, " ===")
            print(out)
            print("")

def main():
    parser = argparse.ArgumentParser(description="pstat: small utility to read /proc/<pid> info")
    parser.add_argument("pid", help="pid to inspect")
    parser.add_argument("--compare", action="store_true", help="run ps/top/pidstat and show outputs for comparison")
    args = parser.parse_args()
    print_summary(args.pid, args.compare)

if __name__ == '__main__':
    main()

File: lab3/gr1sub1/Джига_Анна/src/pstat.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <ctype.h>

#define MAX_BUF 1024

void trim(char* str) {
    if (!str) return;
    char* start = str;
    while (*start && isspace(*start)) start++;
    char* end = str + strlen(str) - 1;
    while (end > start && isspace(*end)) end--;
    memmove(str, start, end - start + 1);
    str[end - start + 1] = '\0';
}

int get_proc_value(const char* filename, const char* key, char* result) {
    FILE* f = fopen(filename, "r");
    if (!f) return 0;
    
    char buf[MAX_BUF];
    int found = 0;
    
    while (fgets(buf, sizeof(buf), f)) {
        if (strncmp(buf, key, strlen(key)) == 0) {
            char* val = strchr(buf, ':');
            if (val) {
                strcpy(result, val + 1);
                trim(result);
                found = 1;
                break;
            }
        }
    }
    
    fclose(f);
    return found;
}


int main(int argc, char* argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <pid>\n", argv[0]);
        return 1;
    }
    
    char* pid = argv[1];
    long hz = sysconf(_SC_CLK_TCK);
    char stat_path[MAX_BUF];
    snprintf(stat_path, sizeof(stat_path), "/proc/%s/stat", pid);
    
    FILE* f = fopen(stat_path, "r");
    if (!f) {
        perror("Cannot open stat file");
        return 1;
    }
    
    char line[MAX_BUF];
    fgets(line, sizeof(line), f);
    fclose(f);

    char name[256], state;
    long ppid, utime, stime;
    
    if (sscanf(line, "%*d (%255[^)]) %c %ld %*s %*s %*s %*s %*s %*s %*s %*s %*s %*s %ld %ld",
               name, &state, &ppid, &utime, &stime) != 5) {
        fprintf(stderr, "Failed to parse stat\n");
        return 1;
    }
    
    char status_path[MAX_BUF], io_path[MAX_BUF], smaps_path[MAX_BUF];
    snprintf(status_path, sizeof(status_path), "/proc/%s/status", pid);
    snprintf(io_path, sizeof(io_path), "/proc/%s/io", pid);
    snprintf(smaps_path, sizeof(smaps_path), "/proc/%s/smaps_rollup", pid);
    
    char threads[64] = "N/A";
    char vmrss[64] = "N/A";
    char voluntary[64] = "N/A";
    char nonvoluntary[64] = "N/A";
    char read_bytes[64] = "N/A";
    char write_bytes[64] = "N/A";
    char rssanon[64] = "N/A";
    char rssfile[64] = "N/A";
    
    get_proc_value(status_path, "Threads", threads);
    get_proc_value(status_path, "VmRSS", vmrss);
    get_proc_value(status_path, "voluntary_ctxt_switches", voluntary);
    get_proc_value(status_path, "nonvoluntary_ctxt_switches", nonvoluntary);
    get_proc_value(io_path, "read_bytes", read_bytes);
    get_proc_value(io_path, "write_bytes", write_bytes);
    get_proc_value(smaps_path, "RssAnon", rssanon);
    get_proc_value(smaps_path, "RssFile", rssfile);
    
    printf("--- Process statistics for PID %s ---\n", pid);
    printf("PPid: %ld\n", ppid);
    printf("Threads: %s\n", threads);
    printf("State: %c\n", state);
    printf("utime: %ld ticks\n", utime);
    printf("stime: %ld ticks\n", stime);
    printf("CPU time: %.2f seconds\n", (double)(utime + stime) / hz);
    printf("voluntary_ctxt_switches: %s\n", voluntary);
    printf("nonvoluntary_ctxt_switches: %s\n", nonvoluntary);
    
    if (strcmp(vmrss, "N/A") != 0 && strlen(vmrss) > 0) {
        char* kb_pos = strstr(vmrss, " kB");
        if (kb_pos) *kb_pos = '\0';
        trim(vmrss);
        
        if (strlen(vmrss) > 0) {
            long rss_kb = atol(vmrss);
            printf("VmRSS: %ld kB\n", rss_kb);
            printf("RSS MiB: %.2f MiB\n", rss_kb / 1024.0);
        } else {
            printf("VmRSS: N/A\n");
            printf("RSS MiB: N/A\n");
        }
    } else {
        printf("VmRSS: N/A\n");
        printf("RSS MiB: N/A\n");
    }
    
    printf("RssAnon: %s\n", rssanon);
    printf("RssFile: %s\n", rssfile);
    
    if (strcmp(read_bytes, "N/A") != 0 && strlen(read_bytes) > 0) {
        long long bytes = atoll(read_bytes);
        printf("read_bytes: %lld bytes", bytes);
        if (bytes >= 1024*1024) printf(" (%.2f MiB)", bytes/(1024.0*1024.0));
        else if (bytes >= 1024) printf(" (%.2f KiB)", bytes/1024.0);
        printf("\n");
    } else {
        printf("read_bytes: N/A\n");
    }
    
    if (strcmp(write_bytes, "N/A") != 0 && strlen(write_bytes) > 0) {
        long long bytes = atoll(write_bytes);
        printf("write_bytes: %lld bytes", bytes);
        if (bytes >= 1024*1024) printf(" (%.2f MiB)", bytes/(1024.0*1024.0));
        else if (bytes >= 1024) printf(" (%.2f KiB)", bytes/1024.0);
        printf("\n");
    } else {
        printf("write_bytes: N/A\n");
    }
    
    return 0;
}
